climada/engine/uncertainty/unc_impact.py

```diff
@@ -0,0 +1,309 @@
+                            calc_eai_exp=False,
+                            calc_at_event=False,
+                            pool=None
+                            ):
+        """
+        Computes the impact for each of the parameters set defined in
+        uncertainty.samples.
+
+        By default, the aggregated average annual impact
+        (impact.aai_agg) and the excees impact at return periods rp
+        (imppact.calc_freq_curve(self.rp).impact) is computed.
+        Optionally, eai_exp and at_event is computed (this may require
+        a larger amount of memory if n_samples and/or the number of centroids
+        is large).
+
+        Parameters
+        ----------
+        rp : list(int), optional
+            Return periods in years to be computed.
+            The default is [5, 10, 20, 50, 100, 250].
+        calc_eai_exp : boolean, optional
+            Toggle computation of the impact at each centroid location.
+            The default is False.
+        calc_at_event : boolean, optional
+            Toggle computation of the impact for each event.
+            The default is False.
+        pool : pathos.pools.ProcessPool, optional
+            Pool of CPUs for parralel computations. Default is None.
+            The default is None.
+
+        Raises
+        ------
+        ValueError:
+            If no sampling parameters defined, the distribution cannot
+            be computed.
+
+        """
+
+        if self.sample.empty:
+            raise ValueError("No sample was found. Please create one first"
+                             "using UncImpact.make_sample(N)")
+
+        if rp is None:
+            rp=[5, 10, 20, 50, 100, 250]
+
+        self.rp = rp
+        self.calc_eai_exp = calc_eai_exp
+        self.calc_at_event = calc_at_event
+        
+        start = time.time()
+        one_sample = self.sample.iloc[0:1].iterrows()
+        imp_metrics = map(self._map_impact_calc, one_sample)
+        [aai_agg_list, freq_curve_list,
+         eai_exp_list, at_event_list, tot_value_list] = list(zip(*imp_metrics))
+        elapsed_time = (time.time() - start) 
+        est_com_time = self.est_comp_time(elapsed_time, pool)
+        LOGGER.info(f"\n\nEstimated computation time: {est_com_time}s\n")
+        
+        #Compute impact distributions
+        if pool:
+            LOGGER.info('Using %s CPUs.', pool.ncpus)
+            chunksize = min(self.n_runs // pool.ncpus, 100)
+            imp_metrics = pool.map(self._map_impact_calc,
+                                           self.samples.iterrows(),
+                                           chunsize = chunksize)
+
+        else:
+            imp_metrics = map(self._map_impact_calc, self.sample.iterrows())
+        
+        #Perform the actual computation
+        with log_level(level='ERROR', name_prefix='climada'):
+            [aai_agg_list, freq_curve_list,
+             eai_exp_list, at_event_list,
+             tot_value_list] = list(zip(*imp_metrics))
```

Comment:
This doesnt look like it performs the actual calculation..

Reply:
That is why I added the comment. The `map` returns an iterator. The evaluation of this iterator happens in this line where it is cast into a list.

Reply:
ah. this is very tricky to spot. can you expand the comment in a way you just explained it here?